### ISSUE TITLE
[HOLD for payment 2023-04-26] [$1000] Red dot remains visible on chat after a failed attachment is deleted

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16852]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Click on “Profile” avatar
2. Click on “Preference” link
3. Turn on “Simulate a network failure”
4. Open an existing chat
5. Click on “Actions” icon
6. Click on “Add attachment” link
7. Select a picture of file
8. Wait few seconds
9. Turn off “Simulate a network failure”
10. Click on “Delete” icon on the failed attachment
11. Observe that the failed attachment is deleted, and the corresponding error indicator ‘Red dot’ remains visible on the chat

## Expected Result: 
Red dot shouldn’t remain visible after a filed attachment is deleted
## Actual Result: 
Red dot remains visible after a failed attachment is deleted

## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.93-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/229311733-d3e450eb-e37c-49b9-845e-399efde391b0.mp4

https://user-images.githubusercontent.com/43996225/229311740-f7a68b5f-2489-44e8-97a9-995256351019.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Natnael-Guchima
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1680249331562699

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01365275cfe90a2203</li>
        <li>Upwork Job ID: 1643570634344288256</li>
        <li>Last Price Increase: 2023-04-12</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @bfitzexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
@bfitzexpensify Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by bfitzexpensify** [comment]
Think this is valid for external

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01365275cfe90a2203

**Comment by MelvinBot** [comment]
Current assignee @bfitzexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @0xmiroslav (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @neil-marcellini (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Red dot in LHN (chat) item still shows after deleting the report action that cause the error indicator.

### What is the root cause of that problem?
When we delete a report action, we never clear the errors.
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/libs/actions/Report.js#L706-L712

### What changes do you think we should make in order to solve the problem?

Add `errors: null` to clear the error in optimistic data and `errors: reportAction.errors` in failure data to revert the `errors` just in case the request fails.
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/libs/actions/Report.js#L725-L737

**Comment by bfitzexpensify** [comment]
Not overdue, melv

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Red dot remains visible after a failed attachment is deleted.

### What is the root cause of that problem?
The root cause is [here](https://github.com/Expensify/App/blob/87eca24a445209b3067d922ed882d931ed3bf0ea/src/libs/OptionsListUtils.js#L217) where we're getting the errors of the report. We don't exclude deleted (or pending deleted) report actions here so after the (errored) report action is deleted, it will still show the red dot.

### What changes do you think we should make in order to solve the problem?

We need to add logic below this line https://github.com/Expensify/App/blob/62276baa62ffefbae4ccfc7ed5aacc61a7322737/src/libs/OptionsListUtils.js#L230 to filter only the report actions that are deleted (or pending deleted).

Then use that filtered list to calculate the list of errors of the report.

If we just set the errors to `null` when deleting the report action, it will not work for the case when we delete the report action from another device. This is because the `errors` is stored locally.

### What alternative solutions did you explore? (Optional)
- Another potential solution is to set the `errors` to `null` in optimistic data (of course will need to restore it in case of request failure) when deleting the action. **And** also set the `errors` to `null` when we detect that the report action has been deleted (to make sure it still works when deleting the report action from another device), this can be done by putting the logic inside a `ONYXKEYS.COLLECTION.REPORT_ACTIONS` Onyx listener for example.
- While investigating this issue, I found another issue, not sure if it should be fixed in this GH but just mentioning for the sake of completeness: When there's error with posting the report action in 1st device, now the `pendingAction` is `add`, and then we delete that report action from a 2nd device. The report action will not disappear from the 1st device.
This is because in here https://github.com/Expensify/App/blob/e3597c68d03ad48b50b328daf5cf3dbd269b678d/src/libs/ReportActionsUtils.js#L209 we still show the report action if the report action has any `pendingAction`. In this case it has `pendingAction` is `add` from the initial report action creating attempt so it still shows even though already deleted.
To fix this we need to update the logic so that if the report action is deleted, it will only still show if the `pendingAction` is `delete`.

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by neil-marcellini** [comment]
@0xmiroslav says he's on it

**Comment by neil-marcellini** [comment]
I kind of like [this proposal](https://github.com/Expensify/App/issues/16852#issuecomment-1497311034). It makes sense to remove the errors optimistically and restore them if deleting fails. Thoughts @0xmiroslav? Can you verify that we don't clear the errors in the response or Pusher data from deleting a comment?

**Comment by bernhardoj** [comment]
@0xmiroslav What is that don't work? If we want to handle when the user deleted the chat from other device, we should clear the errors from the Pusher.

**Comment by tienifr** [comment]
Also IMO the errors of an action and whether the action is deleted are not related to each other. So it might not be good to assume an action does not/did not have an error after it's deleted, it causes data loss. If we don't want to show it in the UI it might be better to filter it out.

**Comment by neil-marcellini** [comment]
> If we just set the errors to `null` when deleting the report action, it will not work for the case when we delete the report action from another device.

@tienifr that doesn't make sense to me, because how would we see the report action (message) on another device if it failed to add? From my local testing, the simulate failing network requests toggle only simulates a failure, but the message actually posts. Since there's no consideration for other devices we can go with @bernhardoj's [proposal](https://github.com/Expensify/App/issues/16852#issuecomment-1497311034) and we won't need a backend change. Sound good @0xmiroslav?

**Comment by neil-marcellini** [comment]
True the whole issue is an edge case, but if for some reason adding an attachment fails and the user decides to delete the comment rather than dismissing the error, we should clear the error. It's quite a simple front end change so we should do it.

**Comment by MelvinBot** [comment]
📣 @bernhardoj You have been assigned to this job by @neil-marcellini!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by bernhardoj** [comment]
Thanks! Will open the PR in a few hours.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@0xmiroslav] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@0xmiroslav] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@0xmiroslav] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@bfitzexpensify] Determine if we should create a regression test for this bug.
- [ ] [@0xmiroslav] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@bfitzexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by bfitzexpensify** [comment]
Offers sent out for payment

**Comment by bfitzexpensify** [comment]
I think adding a regression test here is a good idea, perhaps in the [Compose box - Attachment only (Offline Behavior)](https://expensify.testrail.io/index.php?/cases/view/1971011&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229062) test

**Comment by Natnael-Guchima** [comment]
> Offers sent out for payment

Accepted. Thanks 🙏.

**Comment by bfitzexpensify** [comment]
@0xmiroslav please review the items in https://github.com/Expensify/App/issues/16852#issuecomment-1515074270 when you get a chance - thanks!

**Comment by neil-marcellini** [comment]
Bump @0xmiroslav

**Comment by bfitzexpensify** [comment]
Thanks @0xmiroslav. I think we are good to close this one out now.
